service/backend/pchain/indexer/parser.go

```diff
@@ -0,0 +1,349 @@
+	}
+
+	parsedBlock := ParsedBlock{
+		Height:    blk.Height(),
+		BlockID:   blk.ID(),
+		BlockType: fmt.Sprintf("%T", blk),
+		Proposer:  proposer,
+	}
+
+	blockTimestamp := time.Time{}
+
+	switch castBlk := blk.(type) {
+	case *pBlocks.ApricotProposalBlock:
+		errs.Add(p.initializeTx(castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = castBlk.Txs()
+
+		// If the block has an advance time tx, use its timestamp as the block timestamp
+		for _, tx := range parsedBlock.Txs {
+			if att, ok := tx.Unsigned.(*txs.AdvanceTimeTx); ok {
+				blockTimestamp = att.Timestamp()
+				break
+			}
+		}
+	case *pBlocks.BanffProposalBlock:
+		errs.Add(p.initializeTx(castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = castBlk.Txs()
+		parsedBlock.Txs = append(parsedBlock.Txs, castBlk.Transactions...)
+		blockTimestamp = castBlk.Timestamp()
+	case *pBlocks.ApricotAtomicBlock:
+		errs.Add(p.initializeTx(castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = castBlk.Txs()
+	case *pBlocks.ApricotStandardBlock:
+		for _, tx := range castBlk.Transactions {
+			errs.Add(p.initializeTx(tx))
+		}
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = castBlk.Transactions
+
+	case *pBlocks.BanffStandardBlock:
+		for _, tx := range castBlk.Transactions {
+			errs.Add(p.initializeTx(tx))
+		}
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = castBlk.Transactions
+		blockTimestamp = castBlk.Timestamp()
+	case *pBlocks.ApricotAbortBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+	case *pBlocks.BanffAbortBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+		blockTimestamp = castBlk.Timestamp()
+	case *pBlocks.BanffCommitBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+		blockTimestamp = castBlk.Timestamp()
+	case *pBlocks.ApricotCommitBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+	default:
+		errs.Add(fmt.Errorf("no handler exists for block type %T", castBlk))
+	}
+
+	// If no timestamp was found in a given block (pre-Banff) we fallback to proposer timestamp as used by Snowman++
+	// if available.
+	//
+	// For pre-Snowman++, proposer timestamp does not exist either. In that case, fallback to the genesis timestamp.
+	// This is needed as opposed to simply return 0 timestamp as Rosetta validation expects timestamps to be available
+	// after 1/1/2000.
+	if blockTimestamp.IsZero() {
+		if proposer.Timestamp > genesisTimestamp {
+			blockTimestamp = time.Unix(proposer.Timestamp, 0)
+		} else {
+			blockTimestamp = time.Unix(genesisTimestamp, 0)
+		}
+	}
+	parsedBlock.Timestamp = blockTimestamp.UnixMilli()
+
+	return &parsedBlock, errs.Err
+}
+
+func getProposerFromBytes(bytes []byte) (Proposer, []byte, error) {
+	proposer, _, err := proposerBlk.Parse(bytes)
+	if err != nil || proposer == nil {
```

Comment:
@bilgehansahin why is this error silenced? We really should further process bytes if parser errored

Reply:
I have copied this function verbatim from the indexer parsing reference implementation we were provided. I'm not very familiar with the how proposer works. We can handle it differently if you have a suggestion.

Reply:
I looked into this and it is correct. Briefly, we have to handle blocks before and after snowman++ activation. To do that we try parsing a block as proposerVM block and if it fails, we try and a simple pre-snowman++ Pchain block.

In fact I have refactored this for clarity in the [Indexer cleanup PR](https://github.com/ava-labs/avalanche-rosetta/pull/160). Would really appreciate your input there